packages/core/src/directory.ts

```diff
@@ -0,0 +1,263 @@
+import { Org, OrgID, OrgMember, OrgMemberStatus, Group } from "./org";
+import { Server } from "./server";
+import { getIdFromEmail } from "./util";
+import { Auth } from "./auth";
+
+export interface DirectoryUser {
+    externalId?: string;
+    email?: string;
+    name?: string;
+    active?: boolean;
+}
+
+export interface DirectoryGroup {
+    externalId?: string;
+    name: string;
+    members?: DirectoryUser[];
+}
+
+export interface DirectoryGroupChanges {
+    newName?: string;
+    memberIdsToAdd?: string[];
+    memberIdsToRemove?: string[];
+}
+
+export interface DirectorySubscriber {
+    userCreated(user: DirectoryUser, orgId: OrgID): Promise<OrgMember | void>;
+    userUpdated(user: DirectoryUser, orgId: OrgID, userId: string): Promise<OrgMember | void>;
+    userDeleted(user: DirectoryUser, orgId: OrgID, userId: string): Promise<void>;
+
+    groupCreated(group: DirectoryGroup, orgId: OrgID): Promise<Group | void>;
+    groupUpdated(groupChanges: DirectoryGroupChanges, orgId: OrgID, groupId: string): Promise<Group | void>;
+    groupDeleted(group: DirectoryGroup, orgId: OrgID, groupId: string): Promise<void>;
+}
+
+export interface DirectoryProvider {
+    subscribe(sub: DirectorySubscriber): void;
+}
+
+export class DirectorySync implements DirectorySubscriber {
+    constructor(public readonly server: Server, providers: DirectoryProvider[] = []) {
+        for (const provider of providers) {
+            provider.subscribe(this);
+        }
+    }
+
+    async userCreated(user: DirectoryUser, orgId: string) {
+        const org = (orgId && (await this.server.storage.get(Org, orgId))) || null;
+        if (org && org.directory.syncMembers) {
+            const memberExists = org.members.some((member) => member.email === user.email);
+
+            if (memberExists) {
+                return;
+            }
+
+            const newOrgMember = new OrgMember({
+                name: user.name,
+                email: user.email,
+                status: OrgMemberStatus.Provisioned,
+                updated: new Date(),
+            });
+
+            org.members.push(newOrgMember);
+
+            await this.server.updateMetaData(org);
+            await this.server.storage.save(org);
+
+            return newOrgMember;
+        }
+    }
+
+    async userUpdated(user: DirectoryUser, orgId: string, userId: string) {
+        const org = (orgId && (await this.server.storage.get(Org, orgId))) || null;
+        if (org && org.directory.syncMembers) {
+            let existingUser: OrgMember | null = null;
+            for (const member of org.members) {
+                if (existingUser) {
+                    continue;
+                }
+
+                const provisioningId = await getIdFromEmail(member.email);
+
+                if (member.accountId === userId || member.id === userId || provisioningId === userId) {
```

Comment:
I don't like this ambiguity in how we deal with ids here. When I was testing the endpoint it wasn't clear whether the expected user id was the email, the account id or something else. We should decide on a single identifier and use that throughout. The account id is not a good choice for this since it's not guaranteed to be defined yet, so it must be one of:

1. The email, or a hash of it
2. The external id, mapped to the email
3. A random id, mapped to the email

2 and 3 have the advantage that they allow the id to be constant, which should be a hard requirement for a public id anyway.

Reply:
Agreed. I had the impression the `getIdFromEmail()` would match these, so that's what I was going for here. I prefer your idea of generating something else and keeping that map outside of this.